### ISSUE TITLE
MissingMethodException occured when deploying Blazor Server app with 'Trim Unused code'

### ISSUE THREAD

**Issue by yangzhongke** [dotnet/aspnetcore#40033]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I created an Blazor Server application using the default template on .NET 6, and deploy it with the option 'Trim Unused code' as self-contained. Then, I ran the application, and launched the page in browser, and the 'System.MissingMethodException: Cannot dynamically create an instance of type 'Microsoft.AspNetCore.Components.Web.HeadOutlet'. Reason: No parameterless constructor defined.' occurred.
When I checked HeadOutlet class of the generated Microsoft.AspNetCore.Components.Web.dll, I found that the HeadOutlet class is a empty class.
![image](https://user-images.githubusercontent.com/13821363/152800732-9c3218e5-217f-49ca-9723-18bffcacdd8b.png)

If I deployed the application without the option 'Trim Unused code', the generated Microsoft.AspNetCore.Components.Web.dll is  normal, and no exceptions occurred.

Source code: [https://1drv.ms/u/s!AsFi0-jzsGnQawMBCLgbBt2PYGg?e=BviG3k](https://1drv.ms/u/s!AsFi0-jzsGnQawMBCLgbBt2PYGg?e=BviG3k)

### Steps To Reproduce

1) Create a Blazor Server application.
2) Deploy the application with the option 'Trim Unused code' as self-contained.
3) Run the deployed binaries.
![image](https://user-images.githubusercontent.com/13821363/152800683-3915212a-e73f-490f-a517-473074eac6c2.png)

### Exceptions (if any)

fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.MissingMethodException: Cannot dynamically create an instance of type 'Microsoft.AspNetCore.Components.Web.HeadOutlet'. Reason: No parameterless constructor defined.
         at System.RuntimeType.ActivatorCache..ctor(RuntimeType )
         at System.RuntimeType.CreateInstanceDefaultCtor(Boolean , Boolean )
         at System.Activator.CreateInstance(Type type, Boolean nonPublic, Boolean wrapExceptions)
         at System.Activator.CreateInstance(Type type, Boolean nonPublic)
         at System.Activator.CreateInstance(Type type)
         at Microsoft.AspNetCore.Components.DefaultComponentActivator.CreateInstance(Type )
         at Microsoft.AspNetCore.Components.ComponentFactory.InstantiateComponent(IServiceProvider , Type )
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.InstantiateComponent(Type )
         at Microsoft.AspNetCore.Components.Rendering.HtmlRenderer.CreateInitialRenderAsync(Type componentType, ParameterView initialParameters)
         at Microsoft.AspNetCore.Components.Rendering.HtmlRenderer.RenderComponentAsync(Type componentType, ParameterView initialParameters)
         at Microsoft.AspNetCore.Components.Rendering.RendererSynchronizationContext.<>c__11`1.<<InvokeAsync>b__11_0>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.ViewFeatures.StaticComponentRenderer.PrerenderComponentAsync(ParameterView parameters, HttpContext httpContext, Type componentType)
         at Microsoft.AspNetCore.Mvc.ViewFeatures.ComponentRenderer.PrerenderedServerComponentAsync(HttpContext context, ServerComponentInvocationSequence invocationId, Type type, ParameterView parametersCollection)
         at Microsoft.AspNetCore.Mvc.ViewFeatures.ComponentRenderer.RenderComponentAsync(ViewContext viewContext, Type componentType, RenderMode renderMode, Object parameters)
         at Microsoft.AspNetCore.Mvc.TagHelpers.ComponentTagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)
         at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.<RunAsync>g__Awaited|0_0(Task task, TagHelperExecutionContext executionContext, Int32 i, Int32 count)
         at BlazorApp1.Pages.Pages__Layout.<ExecuteAsync>b__14_0() in C:\Users\cowne\source\repos\BlazorApp1\BlazorApp1\Pages\_Layout.cshtml:line 14
         at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperExecutionContext.SetOutputContentAsync()
         at BlazorApp1.Pages.Pages__Layout.ExecuteAsync()
         at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
         at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
         at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderLayoutAsync(ViewContext context, ViewBufferTextWriter bodyWriter)
         at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
         at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
         at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)

### .NET Version

6.0.101

### Anything else?

Microsoft Visual Studio Professional 2022 (64-bit)  Version 17.0.5

.NET SDK:
 Version:   6.0.101
 Commit:    ef49f6213a

 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.101\

Host (useful for support):
  Version: 6.0.1
  Commit:  3a25a7f1cc

.NET SDKs installed:
  5.0.402 [C:\Program Files\dotnet\sdk]
  6.0.101 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21501.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by pranavkm** [comment]
This is similar to https://github.com/dotnet/aspnetcore/issues/37143, but with Server rendering, not WASM. Could you use https://github.com/dotnet/aspnetcore/issues/37143#issuecomment-931726256 as a workaround to unblock yourself?

**Comment by mkArtakMSFT** [comment]
@yangzhongke can you try out the latest .NET 6.0 SDK release and see whether you can still hit this issue? Thanks

**Comment by yangzhongke** [comment]
> HeadOutlet

Thanks for your reply. However, it still doesn't work.
SDK version: 6.0.4
![image](https://user-images.githubusercontent.com/13821363/163295647-692335e9-b830-4abe-a949-ade75ffc7ba1.png)

![image](https://user-images.githubusercontent.com/13821363/163295579-5ed48261-11e3-4ec9-a348-c887574a435f.png)

**Comment by chrdlx** [comment]
This still happens in .NET 7 Preview 7.

**Comment by hanxiao123** [comment]
> @yangzhongke can you try out the latest .NET 6.0 SDK release and see whether you can still hit this issue? Thanks

yse,latest .NET 6.0 SDK release  can still hit this issue!

**Comment by wizpresso-steve-cy-fan** [comment]
It seems like the trimmer had trimmed away the parameterless constructor due to it is obviously not used statically unless it is under reflection context. This issue is prominent in running with deserialization framework, and with the use of record where you cannot force to create a parameterless constructor I think.

**Comment by danmoseley** [comment]
Cc @eerhardt

**Comment by Jinjinov** [comment]
This also happens in my MAUI Blazor project: https://github.com/Jinjinov/Ididit

I can disable trimming for Linux and Android, but there is no way to disable trimming for iOS and macOS

```
Arg_NoDefCTor, Microsoft.AspNetCore.Components.Web.HeadOutlet
   at System.RuntimeType.CreateInstanceMono(Boolean , Boolean )
   at System.RuntimeType.CreateInstanceDefaultCtor(Boolean , Boolean )
   at System.Activator.CreateInstance(Type type, Boolean nonPublic, Boolean wrapExceptions)
   at System.Activator.CreateInstance(Type type, Boolean nonPublic)
   at System.Activator.CreateInstance(Type type)
   at Blazorise.ComponentActivator.CreateInstance(Type componentType)
   at Microsoft.AspNetCore.Components.ComponentFactory.InstantiateComponent(IServiceProvider , Type )
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.InstantiateComponent(Type )
   at Microsoft.AspNetCore.Components.RenderTree.WebRenderer.AddRootComponent(Type , String )
   at Microsoft.AspNetCore.Components.WebView.Services.WebViewRenderer.AddRootComponent(Type componentType, String domElementSelector)
   at Microsoft.AspNetCore.Components.WebView.WebViewManager.AttachToPageAsync(String baseUrl, String startUrl)
   at Microsoft.AspNetCore.Components.WebView.IpcReceiver.OnMessageReceivedAsync(PageContext pageContext, String message)
   at Microsoft.AspNetCore.Components.WebView.WebViewManager.<>c__DisplayClass18_0.<<MessageReceived>b__0>d.MoveNext()
```

**Comment by Lover2023** [comment]
Well I tink no to change this if so let's talk about it

**Comment by Lover2023** [comment]
I am thinking of not changing it I believe that I shouldn't be change for Burt fuel reasons

**Comment by Mark-Phillipson** [comment]
Using a Blazor Server .NET 7 having the same issue when trying to trim:

```
Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.MissingMethodException: Cannot dynamically create an instance of type 'Microsoft.AspNetCore.Components.Web.HeadOutlet'. Reason: No parameterless constructor defined.
```

**Comment by Jinjinov** [comment]
@Mark-Phillipson you can fix this issue with this:
```
    // The following dependency prevents HeadOutlet from getting trimmed away
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HeadOutlet))]
    protected override void OnAfterRender(bool firstRender)
    {
        if (firstRender)
        {
        }
    }
```

**Comment by davidfowl** [comment]
I don' see how Blazor server can be supported for trimming when most of ASP.NET Core isn't.

**Comment by eerhardt** [comment]
FYI - It looks like we put a similar work around for Blazor WASM here:

https://github.com/dotnet/aspnetcore/blob/1d04387cf3b68636f8ade0118334f87ae5f56f7d/src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHostBuilder.cs#L41-L44

**Comment by rolfbjarne** [comment]
> I can disable trimming for Linux and Android, but there is no way to disable trimming for iOS and macOS

You can do this for iOS and Mac Catalyst:

```xml
<PropertyGroup>
    <MouchLink>None</MtouchLink>
</PropertyGroup>
```

We're updating the error message to suggest this in a future release.

**Comment by javiercn** [comment]
Going to track progress for this topic in the following issue https://github.com/dotnet/aspnetcore/issues/51598 if you are interested, upvote that issue instead.

**Comment by mkArtakMSFT** [comment]
Closing in favor of the issue @javiercn linked above, which we will be prioritizing for .NET 9.
